### PR TITLE
Improve efficiency of array/object serialization in query plans

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Follow-up PR to https://github.com/arangodb/arangodb/pull/17897

* Use more compact and efficient representation for arrays and objects during AQL AST serialization and deserialization. This can help to reduce the size of messages exchanged between coordinator and database servers during query setup, and also reduce the time needed for parsing these messages. This especially helps when there are large bind parameter values that are arrays or objects.

- [ ] :hankey: Bugfix
- [ ] :pizza: New feature
- [x] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [x] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.10: -
  - [ ] Backport for 3.9: -
  - [ ] Backport for 3.8: -

#### Related Information

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/Aql/AqlCallList.cpp [hunk 1]
```diff
@@ -121,9 +115,13 @@ auto AqlCallList::fromVelocyPack(VPackSlice slice) -> ResultT<AqlCallList> {
...
   }
 
-  auto expectedPropertiesFound = std::map<std::string_view, bool>{};
-  expectedPropertiesFound.emplace(StaticStrings::AqlCallListSpecific, false);
-  expectedPropertiesFound.emplace(StaticStrings::AqlCallListDefault, false);
+  // we only have 2 different keys to check. using an std::map requires
+  // dynamic memory allocation and would be wasteful. instead, use a simple
+  // std::array here to get rid of any allocations
+  auto expectedPropertiesFound =
+      std::array<std::pair<std::string_view, bool>, 2>{
```
Comment (MBkkt) [inline arangod/Aql/AqlCallList.cpp:122]: Why do we need bool is it always false?

File: arangod/Aql/SortingGatherExecutor.cpp [hunk 2]
```diff
@@ -358,14 +358,14 @@ auto SortingGatherExecutor::produceRows(typename Fetcher::DataRange& input,
     // First initialize
     auto const callSet = initialize(input, output.getClientCall());
     if (!callSet.empty()) {
-      return {ExecutorState::HASMORE, NoStats{}, callSet};
+      return {ExecutorState::HASMORE, NoStats{}, std::move(callSet)};
```
Comment (MBkkt) [inline arangod/Aql/SortingGatherExecutor.cpp:361]: Move const value is copy

Same in other places here

File: arangod/Aql/AstNode.cpp [hunk 3]
```diff
@@ -799,58 +832,56 @@ void AstNode::sort() {
...
     // specialized compare function that doesn't need to resolve
     // attribute accesses
-    std::sort(
-        members.begin(), members.end(),
-        [](AstNode const* lhs, AstNode const* rhs) {
-          return (arangodb::aql::CompareAstNodes<false>(lhs, rhs, true) < 0);
-        });
+    std::sort(members.begin(), members.end(),
+              [](AstNode const* lhs, AstNode const* rhs) {
+                return (compareAstNodes<false>(lhs, rhs, true) < 0);
```
Comment (MBkkt) [inline arangod/Aql/AstNode.cpp:837]: Nit
Branches unnecessary

File: arangod/Aql/AqlCallList.cpp [hunk 4]
```diff
@@ -177,16 +175,15 @@ auto AqlCallList::fromVelocyPack(VPackSlice slice) -> ResultT<AqlCallList> {
...
-    auto const key = getStringView(keySlice);
+    auto key = it.key.stringView();
 
-    if (auto propIt = expectedPropertiesFound.find(key);
+    if (auto propIt = std::find_if(
+            expectedPropertiesFound.begin(), expectedPropertiesFound.end(),
+            [&key](auto const& epf) { return epf.first == key; });
         ADB_LIKELY(propIt != expectedPropertiesFound.end())) {
       if (ADB_UNLIKELY(propIt->second)) {
+        // should never happen
```
Comment (MBkkt) [inline arangod/Aql/AqlCallList.cpp:185]: I think better to remove such code?

**Review by MBkkt — APPROVED** [review]

File: arangod/Aql/AqlCallList.cpp (hunk 1 above)
Comment (jsteemann) [reply arangod/Aql/AqlCallList.cpp:122]: The boolean values in the array are `false` only initially. They can be updated to `true` in line 191.

File: arangod/Aql/AqlCallList.cpp (hunk 4 above)
Comment (jsteemann) [reply arangod/Aql/AqlCallList.cpp]: Removed

File: arangod/Aql/SortingGatherExecutor.cpp (hunk 2 above)
Comment (jsteemann) [reply arangod/Aql/SortingGatherExecutor.cpp:361]: Fixed

File: arangod/Aql/AqlCallList.cpp (hunk 1 above)
Comment (MBkkt) [reply arangod/Aql/AqlCallList.cpp:122]: Understood

File: arangod/Aql/AqlCallList.cpp [hunk 5]
```diff
@@ -177,20 +175,12 @@ auto AqlCallList::fromVelocyPack(VPackSlice slice) -> ResultT<AqlCallList> {
...
-    }
-    auto const key = getStringView(keySlice);
+    auto key = it.key.stringView();
 
-    if (auto propIt = expectedPropertiesFound.find(key);
+    if (auto propIt = std::find_if(
+            expectedPropertiesFound.begin(), expectedPropertiesFound.end(),
+            [&key](auto const& epf) { return epf.first == key; });
         ADB_LIKELY(propIt != expectedPropertiesFound.end())) {
-      if (ADB_UNLIKELY(propIt->second)) {
```
Comment (MBkkt) [inline arangod/Aql/AqlCallList.cpp:189]: `TRI_ASSERT(!propIt->second)`?

File: arangod/Aql/AstNode.cpp [hunk 6]
```diff
@@ -43,7 +43,9 @@
 #ifdef ARANGODB_ENABLE_MAINTAINER_MODE
 #include <iostream>
 #endif
-#include <unordered_map>
+
+#include <frozen/string.h>
```
Comment (MBkkt) [inline arangod/Aql/AstNode.cpp:47]: Is it really necessary?
```suggestion
```

**Review by MBkkt — APPROVED** [review]

File: arangod/Aql/AstNode.cpp (hunk 6 above)
Comment (jsteemann) [reply arangod/Aql/AstNode.cpp]: Removed

File: arangod/Aql/AqlCallList.cpp (hunk 5 above)
Comment (jsteemann) [reply arangod/Aql/AqlCallList.cpp:189]: Added

**Pull request merged**